File: docs/sliding_window/0076_hard/min_window_substring.py

def find_min_window_substring(source_str: str, target_str: str) -> str:  # LeetCode Q.76.
    tgt_chars_counts: dict[str, int] = dict()

    for char in target_str:
        if char not in tgt_chars_counts.keys():
            tgt_chars_counts.update({char: 0})
        tgt_chars_counts[char] += 1

    unique_tgt_chars_count = len(tgt_chars_counts)

    # Must not exceed unique target chars count during sliding window.
    unique_tgt_chars_coverage = 0

    min_window_left_idx = -1
    min_len = len(source_str) + 1  # Source length + 1: symbol of not found yet.

    left_idx = 0
    for right_idx, char in enumerate(source_str):
        if char in tgt_chars_counts.keys():
            tgt_chars_counts[char] -= 1

            if tgt_chars_counts[char] == 0:  # Becomes covered.
                unique_tgt_chars_coverage += 1

        while unique_tgt_chars_coverage == unique_tgt_chars_count:  # Can shrink window.
            window_len = right_idx + 1 - left_idx

            if window_len < min_len:
                min_len = window_len
                min_window_left_idx = left_idx

            removed_char = source_str[left_idx]
            if removed_char in tgt_chars_counts.keys():
                tgt_chars_counts[removed_char] += 1

                if tgt_chars_counts[removed_char] > 0:  # Becomes uncovered.
                    unique_tgt_chars_coverage -= 1

            left_idx += 1

    if min_len == len(source_str) + 1: return ""

    return source_str[min_window_left_idx: min_window_left_idx + min_len]

File: docs/sliding_window/0076_hard/min_window_substring.cpp
#include <string>
#include <unordered_map>
using namespace std;

string findMinWindowSubstring(string sourceString, string targetString) // LeetCode Q.76.
{
    unordered_map<char, int> tgtCharsCounts;

    for (const auto& character : targetString)
        tgtCharsCounts[character]++;

    int uniqueTgtCharsCount = tgtCharsCounts.size();

    // Must not exceed unique target chars count during sliding window.
    int uniqueTgtCharsCoverage = 0;

    int minWindowLeftIdx = -1;
    int minLen = sourceString.size() + 1; // Source length + 1: symbol of not found yet.

    int leftIdx = 0;
    for (int rightIdx = 0; rightIdx < sourceString.size(); rightIdx++) {
        char character = sourceString[rightIdx];
        if (tgtCharsCounts.find(character) != tgtCharsCounts.end()) {
            tgtCharsCounts[character]--;

            if (tgtCharsCounts[character] == 0) // Becomes covered.
                uniqueTgtCharsCoverage++;
        }

        while (uniqueTgtCharsCoverage == uniqueTgtCharsCount) // Can shrink window.
        {
            int windowLen = rightIdx + 1 - leftIdx;

            if (windowLen < minLen)
                minLen = windowLen, minWindowLeftIdx = leftIdx;

            char removedChar = sourceString[leftIdx];
            if (tgtCharsCounts.find(removedChar) != tgtCharsCounts.end()) {
                tgtCharsCounts[removedChar]++;

                if (tgtCharsCounts[removedChar] > 0) // Becomes uncovered.
                    uniqueTgtCharsCoverage--;
            }

            leftIdx += 1;
        }
    }

    if (minLen == sourceString.size() + 1)
        return "";

    return sourceString.substr(minWindowLeftIdx, minLen);
}
